asllib/Typing.ml

```diff
@@ -3818,21 +3818,28 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
     in
     let new_body, ses_body = try_annotate_block env body in
     let ses = SES.union ses_func_sig @@ SES.remove_locals ses_body in
-    let+ () =
-      if func_version f == V0 || C.fine_grained_side_effects then ok
-      else check_subprogram_purity ~loc:new_body f.qualifier ses
-    in
-    let ses = SES.set_purity_for_subprogram f.qualifier ses in
     let () =
       if false then
         Format.eprintf "@[<v 2>For program %s, I got side-effects:@ %a@]@."
           f.name SES.pp_print ses
     in
+    let qualifier =
+      if C.fine_grained_side_effects then f.qualifier
+      else if func_version f == V0 then
+        if SES.fine_grained_is_pure ses then
+          if SES.max_time_frame ses = TimeFrame.Constant then Some Pure
```

Comment:
I'm wondering whether it would be worth trying to use the coarse-grained analysis here to derive the keyword - could we take e.g. the minimum global purity at this point?